MINOR. Replace Utils::readFileAsString method to read file as stream

soondenana opened the pull request:
The current Utils::readFileAsString method creates a FileChannel and
memory maps file and copies its content to a String and returns it. But
that means that we need to know the size of the file in advance. This
precludes us from reading files whose size is not known in advance, i.e.
any file opened with flag S_IFIFO.

This change updates the method to use stream to read the content of the file.
It has couple of practical advantages:

1. Allows bash process substitution to pass in strings as file. So we can
say `./bin/kafka-reassign-partitions.sh --reassignment-json-file <(echo
"reassignment json")

2. When adding systest for commands that take file, we don't have to
create real physical file. Instead we can just dump the content of the
file on the command line.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
MINOR. Add a Utils method to read file as stream

The current Utils::readFileAsString method creates a FileChannel and
memory maps file and copies its content to a String and returns it. But
that means that we need to know the size of the file in advance. This
precludes us from reading files whose size is not known in advance, i.e.
any file opened with flag S_IFIFO.

This change adds a method to use stream to read the content of the file.
It has couple of practical advantages:

1. Allows bash process substitution to pass in strings as file. So we can
say `./bin/kafka-reassign-partitions.sh --reassignment-json-file <(echo
"reassignment json")

2. When adding systest for commands that take file, we don't have to
create real physical file. Instead we can just dump the content of the
file on the command line.

soondenana commented:
cc @hachikuji, @jsancio , @mumrah

jsancio commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -614,6 +615,24 @@ public static String stackTrace(Throwable e) {
+        StringBuilder fileText = new StringBuilder();
+        try (FileInputStream fileInputStream = new FileInputStream(path)) {
+            int nextChar;
+            while ((nextChar = fileInputStream.read()) != -1) {
```
This does a system call every time. We can use `InputStream`'s `readAllBytes`: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/io/InputStream.java#L389

jsancio replied:
If we do this, we should have an array of `byte`s and can just call `String​(byte[] bytes, Charset charset)`.

soondenana replied:
Unfortunately the call is not in Java 8. Thanks for suggestion, thankfully this is not in critical performance path, and the files we are reading are less than a kilobyte, so it should be fine.

soondenana replied:
Also charset suggestion is cool. Once we drop Java 8 support, we can come back to this and improve it.

ijuma replied:
This is a utility method so we can't say whether it's used in perf critical code or not, right?

soondenana replied:
Current usage is all in *Command classes, but point is valid. In future someone can put this in performance critical path. I have updated based on @mumrah's suggestion to use `Files#readAllBytes` and this method reads all bytes in one go.

Now the `Files#readAllBytes` uses a SeekableByteChannel and gets the size of the file to work, but in tests it seems to be working for a named pipe which isn't seekable. I am going to look into the native code that returns size to check what is going on. If all checks out I will update the PR with commit.

soondenana replied:
So looked into the code and the `size` method simply does fstat and returns the value (so there is no seek). Here is the code:
```
    if (fstat64(fd, &fbuf) < 0)
        return handle(env, -1, "Size failed");
...
    return fbuf.st_size;
```
For the FIFO the value returned is 0 and then the code in `Files#read` allocates a array of 8K size and reads data in chunks of 8K. So we are good with the latest change that @mumrah suggested. Going to post the commit now.

jsancio commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -614,6 +615,24 @@ public static String stackTrace(Throwable e) {
+            while ((nextChar = fileInputStream.read()) != -1) {
+                fileText.append((char) nextChar);
+            }
+        } catch (Exception ex) {
```
I think this can only catch `IOException`s. We can be more explicit.

soondenana replied:
Done.

jsancio reviewed (approved):
LGTM. Some minor comments.

soondenana committed:
Take care of review comments

mumrah commented:
There is also [`Files#readAllBytes`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#readAllBytes-java.nio.file.Path-). I don't think we need to be dealing with InputStreams ourselves here.

And we should probably supply UTF-8 charset rather than relying on system defaults.

soondenana commented:
> There is also [`Files#readAllBytes`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#readAllBytes-java.nio.file.Path-). I don't think we need to be dealing with InputStreams ourselves here.
> 
> And we should probably supply UTF-8 charset rather than relying on system defaults.

Thanks for this, I was looking around and couldn't find it! Done.

soondenana committed:
Update to not fetch one character at a time

ijuma commented:
Do we need two methods? Also, we should always use utf8.

ijuma commented:
I mean, can this implementation replace the existing readFileAsString.

soondenana commented:
> Do we need two methods? Also, we should always use utf8.

You mean two overloaded methods? Regrading charset I was keeping it same as existing `readFileAsString`. Its no fun to have similar methods have different behavior.

However, if we remove existing `readFileAsString` method, then we can also switch to UTF-8 at same time. Currently `readFileAsString` is only used within "*Commands" (and tools/StreamsResetter), so maybe its not too disruptive to get rid of it and replace with new method.

ijuma commented:
Yeah, I'm suggesting we have a single method for reading a file to string that works for all cases.

soondenana committed:
Create only one copy of readFileAsString method

soondenana commented:
> Yeah, I'm suggesting we have a single method for reading a file to string that works for all cases.

Done. Much simpler review now (5 lines of code). Also, now all commands benefit from this, so all can take string at command line, so providing config as a file is optional.

I will update the PR description to reflect this if this version seem fine.

ijuma reviewed (approved):
LGTM

ijuma commented:
retest this please

ijuma commented:
Two jobs passed, 1 failed with 2 unrelated flaky tests:

> org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testReconfigConnector
> kafka.api.AdminClientIntegrationTest.testElectUncleanLeadersForManyPartitions

ijuma commented:
@soondenana I missed the first time around, can we please add a test that failed with the previous implementation and works with this one?

soondenana committed:
Add unit test to test successful reading from named pipe

soondenana commented:
> @soondenana I missed the first time around, can we please add a test that failed with the previous implementation and works with this one?

Done. The unit tests seem to be bit made up, the systests I am working on have real world use case.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -264,6 +264,63 @@ public void testReadBytes() {
 
+    /**
+     * Test to read a simple file as string.
+     */
```
This comment seems redundant given the test name.

soondenana replied:
Removed.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -264,6 +264,63 @@ public void testReadBytes() {
+     */
+    @Test
+    public void testFileAsStringSimpleFile() throws IOException {
+        File tempFile = File.createTempFile("UtilsTestTempFlie", ".tmp");
```
Any reason not to use `TestUtils.tempFile`?

soondenana replied:
Didn't realize that there is already such method. Used it instead.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -264,6 +264,63 @@ public void testReadBytes() {
+                try {
+                    Files.write(fifo.toPath(), testFileContent.getBytes());
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Hmm, what's the goal of this?

soondenana replied:
To put some data in FIFO that will be consumed by the `Utils.readFileAsString` method we are adding in this change. `assertEquals` method below reads this.

Maybe I am missing something here, but the goal is to create fifo, put some data in there and then read it from the other end of it using the new method we have added.

ijuma replied:
I mean why are we printing the stacktrace to sysout. We should either have the test detect exceptions and fail or log the exception, no?

soondenana replied:
Updated to fail the test. Also added a timeout to test in case read/write to fifo block waiting for other end to produce/consume.

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -264,6 +264,63 @@ public void testReadBytes() {
+        // Use the name to create a FIFO in tmp directory
+        String tmpDir = System.getProperty("java.io.tmpdir");
+        String fifoName = "fifo-" + n + ".tmp";
+        File fifo = new File(tmpDir, fifoName);
```
Is there a reason why we have all this naming logic?

soondenana replied:
Need to create a temporary name w/o creating the file on file system (as that will be created by `mkfifo` command). I couldn't find an existing utility to do this.

Another option will be to create a temp file, then delete it and then use that name. Let me know what you think.

ijuma reviewed (commented):
Thanks for the updates. See the questions below.

soondenana committed:
Take care of minor review comments

One comment change and using TestUtils.tempFile() instaed of
File.createTempFile

ijuma commented on clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java:
```diff
@@ -264,6 +264,60 @@ public void testReadBytes() {
+        File fifo = new File(tmpDir, fifoName);
+        Thread producerThread = null;
+        try {
+            Process mkFifoCommand = new ProcessBuilder("mkfifo", fifo.getCanonicalPath()).start();
```
Is this command guaranteed to be available in every system? We may want to make it conditional if not. Or perhaps there's another way to test this that doesn't involve using an external command?

soondenana replied:
This is part of `Gnu coreutils`. So if you can do things like `ls`, `cat` or `echo`, you should be able to run `mkfifo` too. 

Here are list of coreutils command: https://en.wikipedia.org/wiki/List_of_GNU_Core_Utilities_commands

soondenana committed:
Handle exception by failing unit test

The code was priting the message to System.out (argh..) and ignoring
the message. Changed that to fail the test which will terminate
the thread. To make sure that read end doesn't left hanging forever
added a timeout to the test.

soondenana committed:
Propagate IOException from readFileAsString method

The method signature declares IOException as thrown exception, so no
need to catch and throw again. Simplifies the code.

ijuma reviewed (approved):
LGTM

ijuma commented:
We should keep an eye out for the named pipes test. If it causes problems, we should remove it.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -615,21 +613,13 @@ public static String stackTrace(Throwable e) {
-
     public static String readFileAsString(String path) throws IOException {
-        return Utils.readFileAsString(path, Charset.defaultCharset());
+        log.debug("Reading file {}", path);
```
Hmm, why did we do this? I thought we'd have a try/catch block.

soondenana replied:
The method signature was throwing IOException, and implementation wasn't. The only thing try-catch was adding was to log the file path in case things go wrong, so instead added a log statement for that.

soondenana committed:
Rethrow IOException with the updated message

The message contains path of the file that can't be read.

ijuma reviewed (approved):
LGTM
